src/Clay/Geometry.hs

```diff
@@ -42,6 +51,110 @@ maxHeight = key "max-height"
+  negate = error "negate not implemented for AspectRatio"
+
+-- | An 'AspectRatio' can be converted from a 'Rational',
+-- but other operations are not supported.
+instance Fractional AspectRatio where
+  fromRational = AspectRatio
+  recip  = error  "recip not implemented for AspectRatio"
+
+instance Val AspectRatio where
+  value (AspectRatioValue v) = v
+  value (AspectRatio r) = v
+    where v = fromString $ numerator <> "/" <> denominator :: Value
+          numerator = show (R.numerator r)
+          denominator = show (R.denominator r)
+  value (AspectRatioWithFallback (a, b)) = value a <> " " <> value b
+
+-- | Defines the width to height ratio of an element.
+-- At least one of the width or height must be of automatic size,
+-- otherwise the aspect ratio will be ignored.
+--
+-- It can be given a fixed ratio of the width to the height:
+--
+-- >>> renderWith compact [] $ aspectRatio (4%3)
+-- "{aspect-ratio:4/3}"
+--
+-- It can also be the intrinsic aspect ratio for the element:
+--
+-- >>> renderWith compact [] $ aspectRatio auto
+-- "{aspect-ratio:auto}"
+--
+-- It can be told to use the intrinsic aspect ratio for the element,
+-- but to use a fixed ratio while it is unknown or if the element does not have one:
+--
+-- >>> renderWith compact [] $ aspectRatio $ auto `withFallback` (4%3)
+-- "{aspect-ratio:auto 4/3}"
+--
+-- Corresponds to the
+-- [@aspect-ratio@](https://developer.mozilla.org/en-US/docs/Web/CSS/aspect-ratio)
+-- property in CSS.
+aspectRatio :: AspectRatio -> Css
+aspectRatio = key "aspect-ratio"
+
+-- | The aspect ratio of the width to the height for use with 'aspectRatio'.
+--
+-- Note that this is /not/ the same @%@ operator from the "Data.Ratio" module,
+-- although they do both semantically represent ratios.  The same symbol is used
+-- to signify that the return value is a ratio.
+(%) :: Integer -> Integer -> AspectRatio
+(%) m n = fromRational $ (R.%) m n
+
+-- The same as the normal % operator.
+infixl 7 %
+
+-- | A type class for which a type can have a value with another value as a fallback.
+-- Basically, a type class for types which can use 'withFallback'.
+--
+-- 'withFallback' was defined for 'AspectRatio', but this is a type class
+-- because 'withFallback' is a generic name which we may want to reuse
+-- for other types in the future.
+class WithFallback a where
+  -- | Returns a value where one value has another value as a fallback.
+  --
+  -- * For 'AspectRatio', it can be used to specify that the intrinsic aspect
+  --   ratio should be used, but a fixed ratio can be used as a fallback.
+  withFallback :: a -> a -> a
+
+instance WithFallback AspectRatio where
+  withFallback x@(AspectRatioValue "auto") y@(AspectRatio _) =
+    AspectRatioWithFallback (x, y)
+  withFallback x@(AspectRatio _) y@(AspectRatioValue "auto") =
+    AspectRatioWithFallback (x, y)
+  withFallback _ _ =
+    error "Arguments for aspectRatio . withFallback must be auto and a ratio in either order"
```

Comment:
I understand your reasoning here, but I think there shouldn't be a type class with only a single instance. Either there should be useful instances for other types, or `withFallback` should be monomorphic for now. After all, we can always change it into a typeclass later on when we find more types that should have the instance.

Reply:
Sounds reasonable.  I removed the type class.